File: Count Occurences of Anagrams - GFG/count-occurences-of-anagrams.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
//User function template for C++
class Solution{
public:
	int search(string pat, string txt) {
	   unordered_map<char, int> frequencyMap;
	   int i=0;
	   int j=0;
	   int result=0;
	   int n=txt.size();
	   int k=pat.size();
	   for(int i=0;i<k;i++){
	       frequencyMap[pat[i]]++;
	   }
	   int count=frequencyMap.size();
	   while(j<n){
	       if(frequencyMap.find(txt[j])!=frequencyMap.end()){
	           frequencyMap[txt[j]]--;
	           if(frequencyMap[txt[j]]==0){
	               count--;
	           }
	       }
	       if(j-i+1<k){
	           j++;
	       }
	       else if(j-i+1==k){
	           if(count==0){
	               result++;
	           }
	           if(frequencyMap.find(txt[i])!=frequencyMap.end()){
	                frequencyMap[txt[i]]++;
	                if(frequencyMap[txt[i]]==1){
	                    count++;
	                }
	           }
	           i++;
	           j++;
	       }
	   }
	   return result;
	}
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string pat, txt;
        cin >> txt >> pat;
        Solution ob;
        auto ans = ob.search(pat, txt);
        cout << ans << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Count Occurences of Anagrams - GFG/count-occurences-of-anagrams.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    // Driver code
    public static void main(String[] args) throws Exception {
        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while (t-- > 0) {
            String txt = br.readLine().trim();
            String pat = br.readLine().trim();

            int ans = new Solution().search(pat, txt);

            System.out.println(ans);
        }
    }
}
// } Driver Code Ends

class Solution {
    int search(String pat, String txt) {
        int n = txt.length();
        int k = pat.length();
        int count = 0;
        
        int txtFreq[] = new int[26];
        int patFreq[] = new int[26];
        
        for(int i=0; i<k; i++){
            txtFreq[txt.charAt(i) - 'a']++;
            patFreq[pat.charAt(i) - 'a']++;
        }
        
        for(int i=k; i<n; i++){
            if(Arrays.equals(txtFreq, patFreq)){
                count++;
            }
            
            //Remove the leftmost character from the window
            txtFreq[txt.charAt(i-k) - 'a']--;
            
            //Add the rightmost character to the window
            txtFreq[txt.charAt(i) - 'a']++;
        }
        
        if(Arrays.equals(txtFreq, patFreq)){
            count++;
        }
        
        return count;
    }
}